enhancements/installer/pre-flight-validations.md

```diff
@@ -0,0 +1,168 @@
+---
+title: pre-flight-validations
+authors:
+  - "@mandre"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-05-18
+last-updated: 2020-12-01
+status: implementable
+---
+
+# Pre-flight validations
+
+## Release Signoff Checklist
+
+- [v] Enhancement is `implementable`
+- [v] Design details are appropriately documented from clear requirements
+- [v] Test plan is defined
+- [x] Graduation criteria for dev preview, tech preview, GA
+- [x] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+One of the guiding principles of OCP 4 is that the installation should always
+succeed. This goal is relatively easy to implement for public cloud platforms
+where we can make assumptions about services being available, or performance
+meeting requirements, however this is not the case with private clouds where
+each cloud is unique.
+
+It is currently not possible to tell with confidence if an installation will be
+successful or not for the following platforms:
+- BareMetal
+- OpenStack
+- oVirt
+- vSphere
+
+For this purpose, we propose to implement a framework allowing the installer to
+run pre-flight validations in order to certify that all pre-requisites are met
+to successfully install OpenShift in the selected environment.
+
+## Motivation
+
+This section is for explicitly listing the motivation, goals and non-goals of
+this proposal. Describe why the change is important and the benefits to users.
+
+### Goals
+
+By having an automated way to identify potential issues early, before the
+deployment even started, we want to reduce wasted time and resources, customer
+escalations, and improve the perception of OpenShift deployments on-premise.
+
+The installer already performs some validations:
+- checks all required fields are set and that the data is in the right format
+- some basic validation, such as networks do not overlap
+
+However, this doesn't check that the environment is suitable to install OpenShift:
+- pull secret is valid to fetch the container images
+- the tenant has adequate quota and the flavors' specifications are within the
+  recommended ranges.
+- for user-provided networks, check the subnets have a DHCP server and valid
+  DNS to reach the cloud's endpoints
+- required cloud services are available
+- required storage performance
+
+The pre-flight validations should not alter the target infrastructure nor leave
+behind any new resource.
```

Comment:
If cloud providers are in scope consider that EC2 in a region isn't usable until you try to provision an instance.  This initial provision will fail and region has to be approved for use by AWS.  Depending on the support plan and the region this can take minutes or days.  I am not aware that an API to check this has been added.  Are "checks" like this in scope where the implementation is to try creating some infra then tearing it down if successful?